### PR TITLE
[#3601] feat(trino-connector):  Support automatically loading catalog to Trino by Trino SQL command

### PR DESCRIPTION
**Pull request by diqiu50**
### What changes were proposed in this pull request?

1、Support automatically loading catalog to Trino by using CREATE CATALOG command
2、Support drop catalog
3、Support update catalog

### Why are the changes needed?

Fix: #3601

### Does this PR introduce _any_ user-facing change?

NO

### How was this patch tested?

NO

### CODE REVIEW

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+      throw new TrinoException(GRAVITINO_RUNTIME_ERROR, message, e);
+    }
+  }
+
+  boolean checkCatalogExist(String name) {
+    String showCatalogCommand = String.format("SHOW CATALOGS like '%s'", name);
+    Exception failedException = null;
+    try {
+      int retries = EXECUTE_QUERY_MAX_RETRIES;
+      while (retries-- > 0) {
```
Comment (mchades) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java:163]: Can use `Awaitility`?

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    }
+  }
+
+  void executeSql(String sql) {
+    try {
+      int retries = EXECUTE_QUERY_MAX_RETRIES;
+      Exception failedException = null;
+      while (retries-- > 0) {
+        try (Statement statement = connection.createStatement()) {
+          // check the catalog is already created
```
Comment (mchades) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java:193]: redundant comment

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+ * This class dynamically register the Catalog managed by Gravitino into Trino using Trino CREATE
+ * CATALOG statement. It allows the catalog to be used in Trino like a regular Trino catalog.
+ */
+public class CatalogRegister {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CatalogRegister.class);
+
+  private static final int MIN_TRINO_SPI_VERSION = 435;
+  private static final int EXECUTE_QUERY_MAX_RETRIES = 6;
+  private static final int EXECUTE_QUERY_BACKOFF_TIME = 5;
```
Comment (mchades) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java:42]: Please add the time units to the parameter name

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+        "CREATE CATALOG %s USING gravitino WITH ( \"%s\" = 'true', \"%s\" = '%s', %s)",
+        name,
+        GRAVITINO_DYNAMIC_CONNECTOR,
+        GRAVITINO_DYNAMIC_CONNECTOR_CATALOG_CONFIG,
+        GravitinoCatalog.toJson(gravitinoCatalog),
+        config.toCatalogConfig());
+  }
+
+  private String generateDropCatalogCommand(String name) throws Exception {
+    return String.format("DROP CATALOG %s", name);
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java:122]: I am trying to use double quotes around the catalog name. I've found some bugs and will be working on them.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+      throw new TrinoException(
+          GRAVITINO_MISSING_CONFIG,
+          "Error config for Trino catalog store directory, file not found");
+    }
+  }
+
+  private String generateCreateCatalogCommand(String name, GravitinoCatalog gravitinoCatalog)
+      throws Exception {
+    return String.format(
+        "CREATE CATALOG %s USING gravitino WITH ( \"%s\" = 'true', \"%s\" = '%s', %s)",
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java:113]: The same

**Comment by mchades** [comment]
@yuqi1129 overall LGTM, except the retry code, do you want to take another look?

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+
+  void executeSql(String sql) {
+    try {
+      int retries = EXECUTE_QUERY_MAX_RETRIES;
+      Exception failedException = null;
+      while (retries-- > 0) {
+        try (Statement statement = connection.createStatement()) {
+          // check the catalog is already created
+          statement.execute(sql);
+          return;
```
Comment (yuqi1129) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java:197]: Unnecessary return.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+      throw failedException;
+    } catch (Exception e) {
+      throw new TrinoException(GRAVITINO_RUNTIME_ERROR, "Failed to execute query: " + sql, e);
+    }
+  }
+
+  void unResisterCatalog(String name) {
+    try {
+      if (!checkCatalogExist(name)) {
+        return;
```
Comment (yuqi1129) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java:213]: Better to add some log with warning level

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+          Thread.sleep(EXECUTE_QUERY_BACKOFF_TIME_SECOND * 1000);
+        }
+      }
+      throw failedException;
+    } catch (Exception e) {
+      throw new TrinoException(GRAVITINO_RUNTIME_ERROR, "Failed to execute query: " + sql, e);
+    }
+  }
+
+  void unResisterCatalog(String name) {
```
Comment (yuqi1129) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java:210]: unResisterCatalog -> unresisterCatalog

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+        GRAVITINO_DYNAMIC_CONNECTOR_CATALOG_CONFIG,
+        GravitinoCatalog.toJson(gravitinoCatalog),
+        config.toCatalogConfig());
+  }
+
+  private String generateDropCatalogCommand(String name) throws Exception {
+    return String.format("DROP CATALOG %s", name);
+  }
+
+  void registerCatalog(String name, GravitinoCatalog catalog) {
```
Comment (yuqi1129) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java:127]: Is the method marked as a package that can be accessed for unit tests?

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java (hunk 1 above)
Comment (diqiu50) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java:165]: I don't want to invoke another thread. It has almost executed successfully.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java (hunk 6 above)
Comment (diqiu50) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogRegister.java:197]: If there is no return, the function will throw an exception.

**Review by yuqi1129 — APPROVED** [review]
LGTM

**Pull request merged**